source/test_env.cpp

```diff
@@ -14,115 +14,264 @@
  * limitations under the License.
  */
 
+#include <cctype>
+#include <cstdio>
 #include "mbed-drivers/test_env.h"
 
-// Const strings used in test_end
-const char* TEST_ENV_START = "start";
+// Generic test suite transport protocol keys
+const char* TEST_ENV_END = "end";
+const char* TEST_ENV_EXIT = "__exit";
+const char* TEST_ENV_SYNC = "__sync";
+const char* TEST_ENV_TIMEOUT = "__timeout";
+const char* TEST_ENV_HOST_TEST_NAME = "__host_test_name";
+// Test suite success code strings
 const char* TEST_ENV_SUCCESS = "success";
 const char* TEST_ENV_FAILURE = "failure";
-const char* TEST_ENV_MEASURE = "measure";
-const char* TEST_ENV_END = "end";
+// Test case transport protocol start/finish keys
+const char* TEST_ENV_TESTCASE_START = "__testcase_start";
+const char* TEST_ENV_TESTCASE_FINISH = "__testcase_finish";
+// Code Coverage (LCOV)  transport protocol keys
+const char* TEST_ENV_LCOV_START = "__coverage_start";
 
-/* prototype */
+// LCOV support
 extern "C"
 void gcov_exit(void);
 #ifdef YOTTA_CFG_DEBUG_OPTIONS_COVERAGE
 bool coverage_report = false;
+
+void notify_coverage_start(const char *path) {
+    printf("{{%s;%s;", TEST_ENV_LCOV_START, path);
+}
+
+void notify_coverage_end() {
+    printf("}}" NL);
+}
+
+// notify_coverage_start() PAYLOAD notify_coverage_end()
+// ((coverage_start;path;PAYLOAD}}
+
 #endif
 
-static void led_blink(PinName led, float delay)
-{
-    if (led != NC) {
-        DigitalOut myled(led);
-        while (1) {
-            myled = !myled;
-            wait(delay);
-        }
+void greentea_send_kv(const char *key, const char *val) {
+    if (key && val) {
+        printf("{{%s;%s}}" NL, key, val);
     }
-    while(1);
 }
 
-void notify_start()
-{
-    printf("{{%s}}" NL, TEST_ENV_START);
+void greentea_send_kv(const char *key, const int val) {
+    if (key) {
+        printf("{{%s;%d}}" NL, key, val);
+    }
 }
 
-void notify_performance_coefficient(const char* measurement_name, const int value)
-{
-    printf("{{%s;%s;%d}}" RCNL, TEST_ENV_MEASURE, measurement_name, value);
+void greentea_send_kv(const char *key, const char *val, const int success) {
+    if (key) {
+        printf("{{%s;%s;%d}}" NL, key, val, success);
+    }
 }
 
-void notify_performance_coefficient(const char* measurement_name, const unsigned int value)
-{
-    printf("{{%s;%s;%u}}" RCNL, TEST_ENV_MEASURE, measurement_name, value);
+void greentea_send_kv(const char *key) {
+    if (key) {
+        printf("{{%s;%d}}" NL, key, 0);
+    }
 }
 
-void notify_performance_coefficient(const char* measurement_name, const double value)
-{
-    printf("{{%s;%s;%f}}" RCNL, TEST_ENV_MEASURE, measurement_name, value);
+void notify_start() {
+    // Sync preamble: "{{__sync;0dad4a9d-59a3-4aec-810d-d5fb09d852c1}}"
+    // Example value of sync_uuid == "0dad4a9d-59a3-4aec-810d-d5fb09d852c1"
+	char _key[8] = {0};
+	char _value[48] = {0};
+	greentea_parse_kv(_key, _value, sizeof(_key), sizeof(_value));
+    greentea_send_kv(_key, _value);
```

Comment:
Maybe we can make this more robust, by placing this in a while loop until `_key` is equal to `__sync`?